### PR TITLE
Check dollar signs in one-line code-blocks

### PR DESCRIPTION
**Pull request by AlexWayfer**
Offense them, dollar signs are not required.

Resolve GH-230

### CODE REVIEW

File: packages/remark-lint-no-shell-dollars/index.js [hunk 1]
```diff
@@ -81,11 +80,13 @@ function noShellDollars(tree, file) {
...
     if (!generated(node) && node.lang && flags.indexOf(node.lang) !== -1) {
-      lines = node.value.split('\n')
+      lines = node.value.split('\n').filter(function (line) {
+        return line
+      })
       length = lines.length
       index = -1
 
-      if (length <= 1) {
+      if (length === 0) {
```
Comment (ChristianMurphy) [inline packages/remark-lint-no-shell-dollars/index.js:89]: Why did this need to change?

File: packages/remark-lint-no-shell-dollars/readme.md [hunk 2]
```diff
@@ -31,13 +31,11 @@ This rule is included in the following presets:
 ````markdown
 ```sh
 echo a
-echo a > file
```
Comment (ChristianMurphy) [inline packages/remark-lint-no-shell-dollars/readme.md:34]: these example should still work, why were the removed?

File: packages/remark-lint-no-shell-dollars/index.js [hunk 3]
```diff
@@ -59,6 +57,7 @@ var reason = 'Do not use dollar signs before shell commands'
 // See: <https://github.com/github/linguist/blob/40992ba/lib/linguist/languages.yml#L4984>
 var flags = [
   'sh',
+  'shell',
```
Comment (ChristianMurphy) [inline packages/remark-lint-no-shell-dollars/index.js:60]: this seems to be an unrelated change?

File: packages/remark-lint-no-shell-dollars/index.js (hunk 1 above)
Comment (AlexWayfer) [reply packages/remark-lint-no-shell-dollars/index.js:89]: This is main change: https://github.com/remarkjs/remark-lint/issues/230#issuecomment-616097974

File: packages/remark-lint-no-shell-dollars/index.js (hunk 3 above)
Comment (AlexWayfer) [reply packages/remark-lint-no-shell-dollars/index.js:60]: Yes, I can undo this, if you want. Looks weird for me to make a new branch, a new commit, a new PR just for one word.

File: packages/remark-lint-no-shell-dollars/index.js (hunk 1 above)
Comment (ChristianMurphy) [reply packages/remark-lint-no-shell-dollars/index.js:89]: Gotcha, stylistic thought, feel free to disagree.
I tend to prefer boundary checks like `length < 1` over equality checks `length === 0`

Comment (AlexWayfer) [reply packages/remark-lint-no-shell-dollars/index.js:89]: > Gotcha, stylistic thought, feel free to disagree.
> I tend to prefer boundary checks like `length < 1` over equality checks `length === 0`

Why not `length <= 0`? It's difficult to argument. I guess we should not cover negative numbers (undefined behavior).

Comment (ChristianMurphy) [reply packages/remark-lint-no-shell-dollars/index.js:89]: `length <= 0` is also a boundary check.
`<` is ever so slightly more concise, but both can work.

> I guess we should not cover negative numbers

Why not?
It it's more concise, matches more with the previous logic (also a boundary check), and doesn't add complexity.

Comment (AlexWayfer) [reply packages/remark-lint-no-shell-dollars/index.js:89]: > Why not?

Please, read the following text in the parentheses. You've cut the answer from the quote.

> It it's more concise

What? What do you mean? Like "it's 2 chars shorter"? I don't think it matters.

> matches more with the previous logic (also a boundary check)

Please, describe (with examples if possible).

> and doesn't add complexity.

But it covers unexpected cases! One more time: when a negative value of `length` is possible? Can I add it to tests? No? So let's leave it.

Comment (AlexWayfer) [reply packages/remark-lint-no-shell-dollars/index.js:89]: Also I mean that `=== 0` (I personally prefer `== 0`) more readable for me, like "it equals zero". In Ruby language there is `length.zero?`. But `< 1` pushes me to thoughts like "wait, is negative value of `length` possible?". The correct answer is "no", as I know, so why we should do that.

File: packages/remark-lint-no-shell-dollars/index.js (hunk 3 above)
Comment (AlexWayfer) [reply packages/remark-lint-no-shell-dollars/index.js:60]: @ChristianMurphy, so?

File: packages/remark-lint-no-shell-dollars/readme.md (hunk 2 above)
Comment (AlexWayfer) [reply packages/remark-lint-no-shell-dollars/readme.md:34]: Thank you, specs fixed: deletions reverted, new ones have been added.

File: packages/remark-lint-no-shell-dollars/index.js (hunk 3 above)
Comment (ChristianMurphy) [reply packages/remark-lint-no-shell-dollars/index.js:71]: Please split this into it's own PR, thanks!

File: packages/remark-lint-no-shell-dollars/index.js (hunk 1 above)
Comment (ChristianMurphy) [reply packages/remark-lint-no-shell-dollars/index.js:100]: > > matches more with the previous logic (also a boundary check)
>
> Please, describe (with examples if possible).

`<=` is a boundary check, looking at the diff, the previous code was:
```javascript
if (length <= 1) {
```

> But it covers unexpected cases! One more time: when a negative value of `length` is possible?
> The correct answer is "no", as I know, so why we should do that.

It is unlikely, but possible.
It would 

> But `< 1` pushes me to thoughts like "wait, is negative value of length possible?"

An alternative perspective, coming from more a C background.
It's better to do a boundary check when possible, to prevent the possibility of an underflow.

---

@wooorm any preference on style here?

File: packages/remark-lint-no-shell-dollars/index.js (hunk 3 above)
Comment (wooorm) [reply packages/remark-lint-no-shell-dollars/index.js:71]: Hmm. I don’t think that’s a good idea. The comment above the list contains an URL to what GitHub sees as shell code, which matches the current list. I don’t think we should deviate from that.

File: packages/remark-lint-no-shell-dollars/index.js [hunk 4]
```diff
@@ -59,6 +68,7 @@ var reason = 'Do not use dollar signs before shell commands'
...
```
Comment (wooorm) [inline packages/remark-lint-no-shell-dollars/index.js:71]:
```suggestion
```

File: packages/remark-lint-no-shell-dollars/index.js (hunk 1 above)
Comment (wooorm) [reply packages/remark-lint-no-shell-dollars/index.js:100]: I think this, `=== 0`, is fine!

File: packages/remark-lint-no-shell-dollars/index.js (hunk 3 above)
Comment (AlexWayfer) [reply packages/remark-lint-no-shell-dollars/index.js:71]: > Hmm. I don’t think that’s a good idea. The comment above the list contains an URL to what GitHub sees as shell code, which matches the current list. I don’t think we should deviate from that.

Thank you, I didn't notice that. But it's strange: GitHub catches `shell` well, you can check it yourself. I don't know why, and I don't want to dive into `linguist` sources, but I guess it takes root key (`Shell:` for our case) as option too. **Or** GitHub actually uses an another version of `linguist`.

> Please split this into its own PR, thanks!

OK, removed, sorry.

File: packages/remark-lint-no-shell-dollars/index.js (hunk 4 above)
Comment (AlexWayfer) [reply packages/remark-lint-no-shell-dollars/index.js]: Done, thank you.

File: packages/remark-lint-no-shell-dollars/index.js (hunk 1 above)
Comment (AlexWayfer) [reply packages/remark-lint-no-shell-dollars/index.js:99]: I'm going to resolve this, thank you all. Any other discussions are welcome for me.

**Comment by AlexWayfer** [comment]
All discussions are resolved for now, PR is ready to merge I guess.

**Comment by wooorm** [comment]
Released!

**Comment by wooorm** [comment]
Thanks @AlexWayfer!
